### PR TITLE
Added feature #9102 (awaiting method on Semaphore, that returns number of fibers waiting for permits)

### PR DESCRIPTION
**Pull request by caeus**

### CODE REVIEW

**Review by tmccombs — APPROVED** [review]

File: .jvmopts [hunk 1]
```diff
@@ -1 +1,5 @@
 -Dcats.effect.stackTracingMode=full
+-Dfile.encoding=UTF8
+-Xms265M
+-Xmx12G
+-XX:+UseG1GC
```
Comment (kyri-petrou) [inline .jvmopts:5]: Let's revert this one. You can adjust the `.sbtopts` file which is not git tracked and set adequate values depending on your platform / machine

File: core/shared/src/main/scala/zio/Semaphore.scala [hunk 2]
```diff
@@ -153,6 +158,10 @@ object Semaphore {
 
           ref.modify(loop(n, _, ZIO.unit)).flatten
         }
+        def awaiting(implicit trace: Trace): UIO[Long] = ref.get.map {
+          case Left(queue) => queue.size
+          case Right(_)    => 0L
+        }
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/Semaphore.scala:164]: You can avoid a flatMap by using getWith

```suggestion
        def awaiting(implicit trace: Trace): UIO[Long] = ref.get.getWith {
          case Left(queue) => Exit.succeed(queue.size)
          case Right(_)    => Exit.succeed(0L)
        }
```

File: core/shared/src/main/scala/zio/Semaphore.scala [hunk 3]
```diff
@@ -65,6 +65,11 @@ sealed trait Semaphore extends Serializable {
    * permits and releasing them when the scope is closed.
    */
   def withPermitsScoped(n: Long)(implicit trace: Trace): ZIO[Scope, Nothing, Unit]
+
+  /**
+   * Returns the number of fibers awaiting for permits.
+   */
+  def awaiting(implicit trace: Trace): UIO[Long]
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/Semaphore.scala:72]: Unfortunately this isn't binary compatible. You need to use a stub implementation for it and override it below

**Comment by ghostdogpr** [comment]
Done in https://github.com/zio/zio/pull/9212

**Pull request closed without merging**